ui/raidboss/emulator/data/Combatant.js

```diff
@@ -0,0 +1,76 @@
+'use strict';
+
+class Combatant {
+  constructor(id, name) {
+    this.id = id;
+    this.name = name.split('(')[0];
+    this.states = {};
+    this.significantStates = [];
+    this.latesttimestamp = null;
+    this.job = null;
+    this.level = null;
+  }
+
+  pushState(timestamp, State) {
+    this.states[timestamp] = State;
+    this.latesttimestamp = timestamp;
+    if (!this.significantStates.includes(timestamp))
+      this.significantStates.push(timestamp);
+  }
+
+  nextSignificantState(timestamp) {
+    // Shortcut out if this is significant or if there's no higher significant state
+    let index = this.significantStates.indexOf(timestamp);
+    let lastSignificantStateindex = this.significantStates.length - 1;
+    if (index >= 0 && index < lastSignificantStateindex)
+      return this.states[this.significantStates[index + 1]];
+    else if (index === lastSignificantStateindex ||
+        timestamp > this.significantStates[lastSignificantStateindex])
+      return this.states[this.significantStates[lastSignificantStateindex]];
+
+    let i = 0;
+    for (; i < this.significantStates.length; ++i) {
+      if (this.significantStates[i] > timestamp)
+        return this.states[this.significantStates[i]];
+    }
+    return this.states[this.significantStates[i]];
```

Comment:
I'm not sure I understand this loop variable.  Is this just a synonym for "return;" or do you mean to be returning something else?

Reply:
Unless I've miscalculated this should return the last significant state, though I haven't put a breakpoint in and double checked that yet.

Reply:
I guess I'm just reading this as, if you get to the end of the loop and don't find a significant state, then `i === this.significantStates.length` and so `return this.states[this.significantStates[this.significantStates.length]];` is a phrase that means `return;`

Reply:
Yeah, that's probably the case. Like I said, I didn't sit down with a breakpoint in the code to check this. It never threw any error/exception when it hit. I'll just modify the end return to return the last significant state instead.